File: runner/logging.py
#!/usr/bin/env python3
"""Safe task log writer."""

from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path

DEFAULT_LOGS_DIR = Path("/data/clawbrain/logs")


def _ensure_inside(base_dir: Path, target: Path) -> None:
    resolved_base = base_dir.resolve()
    resolved_target = target.resolve()
    if not resolved_target.is_relative_to(resolved_base):
        raise ValueError(f"path escapes log root: {resolved_target}")


def resolve_log_path(task_id: str, logs_dir: str | Path | None = None) -> Path:
    task = (task_id or "").strip()
    if not task:
        raise ValueError("task_id must be non-empty")
    if "/" in task or "\\" in task:
        raise ValueError("task_id must not contain path separators")

    root = Path(logs_dir or DEFAULT_LOGS_DIR).resolve()
    root.mkdir(parents=True, exist_ok=True)
    log_path = (root / f"{task}.log").resolve()
    _ensure_inside(root, log_path)
    return log_path


def append_log(log_path: str | Path, message: str, *, logs_dir: str | Path | None = None) -> None:
    path = Path(log_path).resolve()
    root = Path(logs_dir or DEFAULT_LOGS_DIR).resolve()
    root.mkdir(parents=True, exist_ok=True)
    _ensure_inside(root, path)

    timestamp = datetime.now(timezone.utc).isoformat()
    line = f"[{timestamp}] {message.rstrip()}\n"
    with path.open("a", encoding="utf-8") as handle:
        handle.write(line)

File: src/logging.ts
import fs from "node:fs";
import path from "node:path";

const DEFAULT_LOGS_DIR = "/data/clawbrain/logs";

function isPathInside(basePath: string, targetPath: string, allowEqual: boolean): boolean {
  const relative = path.relative(basePath, targetPath);
  if (!relative) {
    return allowEqual;
  }
  return !relative.startsWith("..") && !path.isAbsolute(relative);
}

function ensureInside(baseDir: string, target: string): void {
  const resolvedBase = path.resolve(baseDir);
  const resolvedTarget = path.resolve(target);
  if (!isPathInside(resolvedBase, resolvedTarget, true)) {
    throw new Error(`path escapes log root: ${resolvedTarget}`);
  }
}

export function resolveLogPath(taskId: string, logsDir = DEFAULT_LOGS_DIR): string {
  const task = (taskId ?? "").trim();
  if (!task) {
    throw new Error("task_id must be non-empty");
  }
  if (task.includes("/") || task.includes("\\")) {
    throw new Error("task_id must not contain path separators");
  }

  const root = path.resolve(logsDir);
  fs.mkdirSync(root, { recursive: true });
  const logPath = path.resolve(path.join(root, `${task}.log`));
  ensureInside(root, logPath);
  return logPath;
}

export function appendLog(logPath: string, message: string, logsDir = DEFAULT_LOGS_DIR): void {
  const targetPath = path.resolve(logPath);
  const root = path.resolve(logsDir);
  fs.mkdirSync(root, { recursive: true });
  ensureInside(root, targetPath);

  const line = `[${new Date().toISOString()}] ${message.trimEnd()}\n`;
  fs.appendFileSync(targetPath, line, { encoding: "utf-8" });
}
